[DOCS] Rewrite of sizing your shards-rebase (#124444)
elastic/elasticsearch: docs/reference/how-to/size-your-shards.asciidoc

[[size-your-shards]]
== Size your shards
[discrete]
[[what-is-a-shard]]
=== What is a shard?

A shard is a basic unit of storage in {es}. Every index is divided into one or more shards to help distribute data and workload across nodes in a cluster. This division allows {es} to handle large datasets and perform operations like searches and indexing efficiently but not without cost. Each index and shard has some overhead and if you divide your data across too many shards then the overhead will degrade performance. Shards play several key roles in {es}:

* *Data Distribution:* Each shard contains a portion of the data from the index. When you add more nodes to your cluster, {es} will spread the shards across the nodes, balancing the workload between them.
* *Replication:* Shards can have replicas which are copies of the original shard. Replicas ensure data availability and improve search performance by allowing multiple nodes to handle requests for that shard.
* *Parallel Processing:* Shards enable {es} to distribute indexing of documents, and process queries in parallel across shards, making ingestion and searches faster and more efficient.

> By effectively using shards, {es} can scale horizontally and provide fault tolerance, ensuring your data is distributed and indexing and searches are processed efficiently.

Reviewer: we should try to keep all of the conceptual info about shards in a single place. we use this page for that in the 8.x docs:

https://www.elastic.co/guide/en/elasticsearch/reference/current/nodes-shards.html

in this doc, I'd just leave a `[shards](link) do XYZ` sentence then dive right into guidelines.

Author: Just a quick thought on this.

I get wanting to avoid duplication, but I think it’s helpful to keep a little bit of the shard context here. I would assume people land on this page without having read the other docs, and this gives them just enough info to get oriented without clicking around. Do we expect the user to go the correct way of looking things up, or expect them to come to this page without context?

Reviewer: we can explain a little about shards, but the purpose of providing the link is so people who feel like they need to know more can go find out more if required. I'd try to keep it to 2-3 sentences. 

we also want to make this page focused, make the docs easier to maintain, and make sure we have a single source of truth for key concepts - linking to one page that dives deep into shards checks all of those boxes.

Author: Ok I made the change, let me know what you think.

Author: @shainaraskas , just a gentle nudge on this

[discrete]
[[sizing-shard-guidelines]]

> === Sizing Shard Guidelines

Reviewer: can drop this heading if you get rid of the conceptual stuff

Reviewer: or change it to "General guidelines"

> Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards (primary or replica), which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures. 
> 
> To strike the right balance, the <<shard-size-recommendation,general guidelines>> are to aim for shard sizes between 10GB and 50GB, keeping the per-shard document count below 200 million. To ensure that each node is working optimally, it's important to distribute shards evenly across nodes. Uneven distribution can cause some nodes to work harder than others, leading to performance degradation and instability. While Elasticsearch automatically balances shards, it’s important to configure your indices with an appropriate number of shards and replicas to facilitate even distribution across nodes.

Reviewer: I'm worried about this being duplicative with the content below, but try something easier to scan:

```suggestion
Balancing the number and size of your shards is important for the performance and stability of an {es} cluster:

* Too many shards can degrade search performance and make the cluster unstable. This is referred to as _oversharding_.
* Very large shards can slow down search operations and prolong recovery times after failures.

To avoid either of these states, implement the following guidelines:

==== General sizing guidelines

* Aim for shard sizes between 10GB and 50GB
* Keep the number of documents on each shard below 200 million

==== Shard distribution guidelines

To ensure that each node is working optimally, distribute shards evenly across nodes. Uneven distribution can cause some nodes to work harder than others, leading to performance degradation and instability. 

While {es} automatically balances shards, you need to configure indices with an appropriate number of shards and replicas to allow for even distribution across nodes.
```

> If you are using <<data-streams>>, each data stream is backed by a sequence of indices, each index potentially having multiple shards. 

Reviewer: why is this important here?

Author: I also tried to remove it, but this is the reason from the previous: https://github.com/elastic/elasticsearch/pull/120346#discussion_r1933605925

> Despite these general guidelines, it is good to develop a tailored <<create-a-sharding-strategy, sharding strategy>> that considers your specific infrastructure, use case, and performance expectations. 

Reviewer: ```suggestion
In addition to these these general guidelines, you should develop a tailored <<create-a-sharding-strategy, sharding strategy>> that considers your specific infrastructure, use case, and performance expectations. 
```
